### ISSUE TITLE
ByteMachine.doTransition takes 50% execution time with latencies seen 160ms

### ISSUE THREAD

**Issue by payalm08** [aws/event-ruler#193]
## To Reproduce
Create Machine instance with 3k/6k/10k rules as below, and run the example query. Results can be captured using jmh. 

Example Rule:

{"data.id":["44422","23232","34325"], // size can be in range of 2-50    
"data.id.333":[{"exists":false}] // size can be in range of 2-50 
}

Example Query: 

{"data.id": ["44422", "23232", "34325"]}. // size can be in range of 50-100 

## Expected behavior
The latency seen is abnormally high. 

## Environment:
Java 8 / Mac OS Sonoma 14.7

**Comment by timbray** [comment]
Question: All the values you're providing look like numbers, but they're actually strings of digits, correct?

The expected latency of Ruler as a function of rule count/shape is (I believe) very similar to that of Quamina, which has a fairly detailed discussion in that repo.  One key aspect is the dependency on the number of unique *keys*. Anyhow, see https://github.com/timbray/quamina?tab=readme-ov-file#flattening-and-matching and especially https://github.com/timbray/quamina?tab=readme-ov-file#matchesforevent-performance

**Comment by payalm08** [comment]
Yes they are strings of digits - does it work differently when compared
with string of alphabets.

**Comment by timbray** [comment]
I don't think that should make a difference, but the value being 1634 vs "1634" might. But apparently that's not what's happening here.

**Comment by payalm08** [comment]
I went through the performance section. For our use case - number of fields
are fixed on pattern. As per the explanation the performance is  O(N) where
for our case N is less than 10.
However, each field is a value array, with variable size array. Is there
any explanation on how the performance vary when array value size changes
both on event and on pattern

**Comment by payalm08** [comment]
Also for our use case when exist false gets added to pattern,  it increases
the number of unique fields on the pattern. So that explains it.

**Comment by fym-rgb** [comment]
Thanks for the detailed report and the follow-up discussion.

I re-ran this issue's shape (rules of `{"data.id": [3 digit-strings], "data.id.333": [{"exists": false}]}`, event with a 1000-element `data.id` array) against **v2.0.1**, where the linear-time `StructuredFinder` engine (introduced in 7677eaf, v1.9.0) became the default for `rulesForJSONEvent` (b0252d2, v2.0.0). Two findings:

**1. The rule-count scaling you saw is match-output size, not machine size.** When the number of *matching* rules is held constant (100) and only non-matching rules are scaled 3k → 6k → 10k, latency is flat on both engines:

| engine | 3k rules | 6k rules | 10k rules | matches |
|---|---|---|---|---|
| StructuredFinder (default, v2.0.0+) | 2588 us | 2330 us | 2274 us | 100 |
| legacy `withStructuredMatching(false)` | 5823 us | 5834 us | 5835 us | 100 |

When rule values are drawn from the same pool as the event's values (so more rules ⇒ more matches: 227 → 477 → 796), latency grows proportionally to the match count — that's the cost of returning more results, which any engine pays.

**2. The default engine is ~2.5x faster on this exact shape.** Same harness, same JVM (JDK 17): ~2.3–2.6 ms vs ~5.8 ms per match call at every rule count.

Combined with your own observation that adding `exists:false` increases the unique-key count (which sets the per-event work, per the Quamina performance notes Tim linked), I believe the remaining per-call cost here is proportional to event-array size and match count, not rule count. Upgrading to v2.0.0+ gets you the faster default engine with no API change.

Closing as resolved by the StructuredFinder default in v2.0.0 (available on the [v2.0.1 release](https://github.com/aws/event-ruler/releases/tag/v2.0.1); note current releases are build-from-source while our Maven Central publishing is being restored). Happy to reopen if you can share a repro where latency grows with rule count at a *fixed* match count.
